feat(receive): Add tests for InputPair constructors Closes #583

Johnosezele opened the pull request:
Add unit tests for the new_p2wpkh and new_p2tr constructors within the `receive` module (payjoin/src/receive/mod.rs). These tests verify that the constructors correctly initialize the InputPair struct, ensuring both the `TxIn` and `psbt::Input` components match expected values.

This improves code quality by confirming constructor behavior and guarding against regressions.

Also addresses compilation errors found during testing:
- Corrected call to new_p2wpkh to include the `sequence` argument.
- Removed assertions checking non-existent fields on `psbt::Input` (e.g., `script_sig`, `tap_control_block`).
- Renamed `tap_leaf_scripts` to the correct field `tap_scripts` in assertions.

benalleng commented on payjoin/src/receive/mod.rs:
```diff
@@ -57,6 +57,80 @@ impl InputPair {
+        Self { txin, psbtin }
+    }
+
+    pub fn new_p2tr(
```
Is there a reason to not deduplicate these constructors as just `new fn input_pair`?

benalleng replied:
Ok I see this referenced in the issue, I guess it would just be nice to have relevant params to give them some reason to be different

Johnosezele replied:
Thanks for the suggestion! Yes, while the constructors currently share similar parameters, keeping them separate allows us to:
- Make the API more explicit about which type of input is being created
- Add Taproot-specific parameters to new_p2tr in the future (like tap internal keys or merkle roots) without breaking changes
- Keep P2TR functionality properly gated behind the v2 feature flag

benalleng replied:
Ok, Its much clearer now the differences between the different descriptors when extending the defaults out.

spacebear21 commented on payjoin/src/receive/mod.rs:
```diff
@@ -57,6 +57,80 @@ impl InputPair {
+        };
+        Self { txin, psbtin }
+    }
+
```
We should offer constructors for at least every input type supported by [`expected_input_weight`](https://github.com/spacebear21/rust-payjoin/blob/master/payjoin/src/psbt/mod.rs#L184), so p2pkh and nested p2wpkh-in-p2sh constructors must be available and tested as well.

Johnosezele replied:
Thanks for the suggestion! I've added constructors for:
-  new_p2pkh() for legacy inputs
-  new_p2sh_p2wpkh() for nested SegWit inputs
I noticed that expected_input_weight() explicitly returns NotSupported for P2WSH inputs. Should we:
-  Keep P2WSH unsupported (current behavior)
-  Add P2WSH support to both expected_input_weight() and create a new new_p2wsh() constructor
What do you think?

spacebear21 commented on payjoin/src/receive/mod.rs:
```diff
@@ -87,3 +161,94 @@ pub(crate) fn parse_payload(
+
+        // Assert: Check the fields of the created InputPair
+        // Check psbtin fields - most should be default/empty
+        assert_eq!(input_pair.psbtin.witness_utxo, Some(dummy_txout));
+        assert_eq!(input_pair.psbtin.non_witness_utxo, None);
+        assert_eq!(input_pair.psbtin.redeem_script, None);
+        assert_eq!(input_pair.psbtin.witness_script, None); // P2TR specific
+        assert!(input_pair.psbtin.partial_sigs.is_empty());
+        assert_eq!(input_pair.psbtin.sighash_type, None);
+        assert_eq!(input_pair.psbtin.final_script_sig, None);
+        assert_eq!(input_pair.psbtin.final_script_witness, None);
+        assert!(input_pair.psbtin.ripemd160_preimages.is_empty());
+        assert!(input_pair.psbtin.sha256_preimages.is_empty());
+        assert!(input_pair.psbtin.hash160_preimages.is_empty());
+        assert!(input_pair.psbtin.hash256_preimages.is_empty()); // P2TR specific check
+        assert_eq!(input_pair.psbtin.tap_key_sig, None);
+        assert!(input_pair.psbtin.tap_script_sigs.is_empty());
+        assert!(input_pair.psbtin.tap_scripts.is_empty());
+        assert!(input_pair.psbtin.tap_key_origins.is_empty()); // P2TR specific check
+        assert_eq!(input_pair.psbtin.tap_internal_key, None); // P2TR specific check
+        assert_eq!(input_pair.psbtin.tap_merkle_root, None); // P2TR specific check
+        assert!(input_pair.psbtin.proprietary.is_empty());
+        assert!(input_pair.psbtin.unknown.is_empty());
+
+        // Check txin field
+        assert_eq!(input_pair.txin.previous_output, dummy_outpoint);
+        assert_eq!(input_pair.txin.script_sig, ScriptBuf::new()); // Should be empty for P2TR
+        assert_eq!(input_pair.txin.sequence, dummy_sequence); // Check sequence default
+        assert!(input_pair.txin.witness.is_empty()); // Witness should be empty initially
```
I don't think checking every default field is useful, I'd rather see us call relevant functions that use `InputPair` such as `expected_input_weight` and ensure they succeed as expected.

Johnosezele replied:
Yeah, true its not useful, I should test behavior instead of implementation. I have added test for [additional_input_weight](https://github.com/spacebear21/rust-payjoin/blob/55a4b7ee15fd31fe88392147adf75fca94ffde50/payjoin/src/receive/v1/mod.rs#L646). I've added test that verifies the relative weights of different input types (P2WPKH vs P2PKH), which is crucial for the fee calculation logic. This test ensures that the weight calculations that feed into fee estimation are working correctly.

Johnosezele replied:
This test verifies that P2WPKH inputs have lower weight than P2PKH inputs, which is one of the key features of witness transactions.

Johnosezele replied:
Hi @spacebear21 I've implemented comprehensive tests for the key privacy and fee-related behaviors in the payjoin protocol. Here's a breakdown of the test coverage:

### Privacy-Based Input Selection Test
The [test_privacy_based_input_selection](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:399:4-482:5) test verifies the core privacy feature of payjoin - avoiding the Unnecessary Input Heuristic (UIH). It:
- Creates two candidate inputs with different amounts (5,000 sats and 50,000 sats)
- Verifies that `try_preserving_privacy` selects an input that maintains privacy
- Tests that the selection logic properly considers input amounts to avoid revealing patterns
- Ensures the receiver's change output is properly handled to maintain privacy

This test is crucial for validating that our implementation effectively prevents blockchain analysis heuristics.

### Fee Rate Enforcement Test
The [test_fee_rate_based_selection](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:484:4-506:5) test focuses on security against probing attacks through fee rate enforcement:
- Sets a minimum fee rate of 2 sat/vB in the proposal
- Verifies that `check_broadcast_suitability` rejects proposals below the minimum fee rate
- Tests the fee rate validation logic in the protocol
- Ensures proper handling of maximum effective fee rates

This helps prevent potential probing attacks by enforcing appropriate fee rates.

### Fee Handling Test
The [test_minimum_fee_requirements](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:508:4-584:5) test validates the complete fee handling flow:
- Tests fee application when contributing inputs
- Verifies fees can be successfully applied within specified min/max ranges
- Uses `finalize_proposal` (the public API) to test the complete proposal flow
- Ensures the receiver can properly cover their own input costs while maintaining protocol security

I've made sure to test both the happy path and edge cases for each of these behaviors. Let me know if you'd like me to add any additional test cases or if you have any suggestions for improving the test coverage for my next pr.

spacebear21 commented on payjoin/src/receive/mod.rs:
```diff
@@ -57,6 +57,80 @@ impl InputPair {
+            partial_sigs: Default::default(),
+            sighash_type: None,
+        };
+        Self { txin, psbtin }
```
It's critical that these InputPair's be validated, which is what the `new` constructor does.  Perhaps `new` should be a private function that gets called in turn by these helper constructors.

Johnosezele replied:
Good point about the [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) constructor! I agree that making it private would better encapsulate the validation logic. Here's what I propose:

1. Make [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) private (remove `pub`) since it's an internal implementation detail
2. Keep the specialized constructors ([new_p2wpkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:88:4-103:5), [new_p2pkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:105:4-120:5), etc.) public as they provide a safer, more guided interface
3. These specialized constructors will internally call the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) function after setting up the correct PSBT inputs

This way:
- We maintain strict validation through the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) function
- Users are guided to use the appropriate constructor for their address type
- We prevent potential misuse by requiring proper UTXO and script setup through the specialized constructors

Would you like me to make these changes in a follow-up PR?

Johnosezele replied:
You make a great point about validation! I initially made [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) private but realized it broke the CLI's usage. After some thought, I have a better approach:

1. Make [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) private (remove `pub`) since it's an internal implementation detail
2. Keep the specialized constructors ([new_p2wpkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:88:4-103:5), [new_p2pkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:105:4-120:5), etc.) public as they provide a safer, more guided interface
3. These specialized constructors will internally call the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) function after setting up the correct PSBT inputs

Benefits of this approach:
- We maintain strict validation through the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) function
- Users are guided to use the appropriate constructor for their address type
- We prevent potential misuse by requiring proper UTXO and script setup through the specialized constructors

I'll make these changes in a follow-up PR.

Johnosezele commented on payjoin/src/receive/mod.rs:
```diff
@@ -57,6 +57,80 @@ impl InputPair {
+            sequence: sequence.unwrap_or_default(),
+            witness: Witness::default(),
+        };
+        let psbtin = psbt::Input {
+            witness_utxo: Some(witness_utxo),
+            non_witness_utxo: None,
+            redeem_script: None,
+            witness_script: None,
+            bip32_derivation: Default::default(),
+            final_script_sig: None,
+            final_script_witness: None,
+            ripemd160_preimages: Default::default(),
+            sha256_preimages: Default::default(),
+            hash160_preimages: Default::default(),
+            hash256_preimages: Default::default(),
+            tap_key_sig: None,
+            tap_script_sigs: Default::default(),
+            tap_scripts: Default::default(),
+            tap_key_origins: Default::default(),
+            tap_internal_key: None,
+            tap_merkle_root: None,
+            proprietary: Default::default(),
+            unknown: Default::default(),
+            partial_sigs: Default::default(),
+            sighash_type: None,
+        };
```
Thanks for the suggestion! You're right - using ..Default::default() is much cleaner and achieves the same result since we're setting everything except witness_utxo to its default value. I've updated both constructor functions to use this pattern. Makes it even more maintainable and less prone to errors if new fields are added to psbt::Input in the future.

Johnosezele commented on payjoin/src/receive/mod.rs:
```diff
@@ -57,6 +57,80 @@ impl InputPair {
+        previous_output: OutPoint,
+        sequence: Option<Sequence>,
+    ) -> Self {
+        let txin = TxIn {
+            previous_output,
+            script_sig: ScriptBuf::new(),
+            sequence: sequence.unwrap_or_default(),
+            witness: Witness::default(),
+        };
```
Fixed too 🫡

Johnosezele replied:
I've implemented them locally and verified all tests pass.

Johnosezele commented:
PR Updated:
- Removed `test_new_p2wpkh` as suggested
- Cleaned up test comments for better readability
- All tests are passing
- Added comprehensive test coverage:
  - `test_privacy_based_input_selection`: Tests UIH avoidance
  - `test_fee_rate_based_selection`: Validates fee rate enforcement
  - `test_minimum_fee_requirements`: Tests end-to-end fee handling

Ready for another review! 👍
